pmlc/dialect/abi/ir/loop_op.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/PatternMatch.h"
+
+#include "pmlc/dialect/abi/ir/dialect.h"
+
+namespace pmlc::dialect::abi {
+
+namespace {
+
+struct DenormalizeConstantsPattern final
+    : public mlir::OpRewritePattern<LoopOp> {
+  using mlir::OpRewritePattern<LoopOp>::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(LoopOp loopOp, mlir::PatternRewriter &rewriter) const final {
+    mlir::OpBuilder::InsertionGuard insertionGuard{rewriter};
+    auto networkOp = loopOp.getInitTerminator();
+    auto networkFieldTypes = loopOp.getAttrOfType<mlir::ArrayAttr>(
+        abi::LoopOp::getNetworkFieldTypesAttrName());
+    mlir::SmallVector<mlir::Attribute, 8> newNetworkFieldTypes;
+    rewriter.startRootUpdate(loopOp);
+    rewriter.setInsertionPointToStart(&loopOp.bodyRegion().front());
+    unsigned idx = 0;
+    for (auto tyAttr : networkFieldTypes) {
+      auto val = networkOp.getOperand(idx);
+      auto definingOp = val.getDefiningOp();
+      if (definingOp && definingOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+        // N.B. Constant-Like operations have no side-effects, exactly one
+        //      result, and no operands.
+        auto *denormOp = definingOp->clone();
+        rewriter.insert(denormOp);
+        auto arg = loopOp.bodyEntryBlock()->getArgument(idx);
+        arg.replaceAllUsesWith(denormOp->getResult(0));
+        loopOp.bodyEntryBlock()->eraseArgument(idx);
+        networkOp.getOperation()->eraseOperand(idx);
```

Comment:
Use `m_Constant` rather than looking for `ConstantLike`

Reply:
Just wondering -- what's wrong with `ConstantLike`?

Reply:
Nothing, it's just that `m_Constant` is probably a much more concise way of matching a constant op. But if it's unsuitable, obviously this way is fine :)